### PR TITLE
Initial commit of pivoted cholesky algorithm from GPyTorch

### PR DESCRIPTION
**Pull request by kunalghosh**
**What is this PR about?**
 fast exact Gaussian processes ([Gardner et.al 2018](https://proceedings.neurips.cc/paper/2018/file/27e8e17134dd7083b050476733207ea1-Paper.pdf)) works on a `modified batch conjugate gradient descent` algorithm which needs a function to return the preconditioning matrix, given a kernel. This PR implements the function to compute the preconditioning matrix.

**Checklist**
+ [√] Explain important implementation details 👆
+ [√] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Are the changes covered by tests and docstrings? (Not yet)

## Major / Breaking Changes
- None

## Bugfixes / New features
- New feature.

## Docs / Maintenance
- Yet to be implemented.

### CODE REVIEW

**Comment by kunalghosh** [comment]
This is a draft pull request for my GSoC Project.

**Comment by kunalghosh** [comment]
This produces the same results as `gpytorch.pivoted_cholesky()`

Test kernel matrices can be generated using the following code which generates random positive semi-definite matrices

```python
import torch 
import gpytorch
import numpy as np

# Test PSD Matrix

N = 10
rank = 5
np.random.seed(1234) # nans with seed 1234
K = np.random.randn(N, N)
K = K @ K.T + N * np.eye(N)
K_torch = torch.from_numpy(K)
```

Now passing the same kernel matrix to the two functions will yield the same output.

```python
L_gpt = gpytorch.pivoted_cholesky(K_torch, rank=rank, error_tol=1e-3)
L_np  = pivoted_cholesky_np_gpt(K, max_iter=rank, error_tol=1e-3)
assert np.allclose(L_gpt, L_np.T) == True, "BUG: The two cholesky decompositions are not same !"

```

**Comment by bwengals** [comment]
Hey @kunalghosh, so sorry again for missing this.  Read through it, it looks good.  Definitely with the torch dependency I think you should move it to `pymc-experimental`.  Once there, tag me for a review and we can get it in ASAP

**Comment by kunalghosh** [comment]
Closing, migrated to pymc_experimental
